package com.devtrack.devtrack_backend.config;

import com.devtrack.devtrack_backend.security.JwtAuthenticationFilter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity
@EnableMethodSecurity
public class SecurityConfig {

    private final JwtAuthenticationFilter jwtAuthenticationFilter;

    public SecurityConfig(
            JwtAuthenticationFilter jwtAuthenticationFilter) {

        this.jwtAuthenticationFilter =
                jwtAuthenticationFilter;
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public SecurityFilterChain securityFilterChain(
            HttpSecurity http) throws Exception {

        http
                .csrf(csrf -> csrf.disable())

                .cors(cors -> {})

                .authorizeHttpRequests(auth -> auth

                        // Public
                        .requestMatchers(
                                "/api/auth/**"
                        ).permitAll()

                        // Only QA
                        .requestMatchers(
                                org.springframework.http.HttpMethod.POST,
                                "/api/issues"
                        ).hasRole("QA")

                        // Issue modification
//                        .requestMatchers(
//                                org.springframework.http.HttpMethod.PUT,
//                                "/api/issues/**"
//                        ).hasAnyRole(
//                                "QA",
//                                "ADMIN"
//                        )

                        .requestMatchers(
                                org.springframework.http.HttpMethod.DELETE,
                                "/api/issues/**"
                        ).hasAnyRole(
                                "QA",
                                "ADMIN"
                        ) .requestMatchers(
                                        org.springframework.http.HttpMethod.DELETE,
                                        "/api/users/**"
                                ).hasAnyRole(
                                        "ADMIN"
                                )



                        // Everything else requires login
                        .anyRequest().authenticated()
                )

                .addFilterBefore(
                        jwtAuthenticationFilter,
                        UsernamePasswordAuthenticationFilter.class
                );

        return http.build();
    }
}
